network/src/lib.rs

```diff
@@ -49,3 +49,76 @@ pub use tokio;
 
 /// Protocol version used by network protocol open
 pub type ProtocolVersion = String;
+
+/// Observe listen port occupancy
+pub fn observe_listen_port_occupancy(
+    _addrs: &[multiaddr::MultiAddr],
+) -> Result<(), std::io::Error> {
+    #[cfg(target_os = "linux")]
+    {
+        use multiaddr::Protocol;
+        use std::net::{IpAddr, SocketAddr, TcpListener, ToSocketAddrs};
+
+        for addr in _addrs {
+            let mut iter = addr.iter().peekable();
+
+            let mut ip_addr: Option<SocketAddr> = None;
+
+            while iter.peek().is_some() {
+                match iter.peek() {
+                    Some(Protocol::Ip4(_))
+                    | Some(Protocol::Ip6(_))
+                    | Some(Protocol::Dns4(_))
+                    | Some(Protocol::Dns6(_)) => (),
+                    _ => {
+                        // ignore is true
+                        let _ignore = iter.next();
+                        continue;
+                    }
+                }
+
+                let proto1 = iter.next().unwrap();
+                let proto2 = if iter.peek().is_some() {
+                    iter.next().unwrap()
+                } else {
+                    break;
+                };
+
+                match (proto1, proto2) {
+                    (Protocol::Ip4(ip), Protocol::Tcp(port)) => {
```

Comment:
suggest to use `multiaddr_to_socketaddr` fn to extract the SocketAddr

Reply:
> suggest to use `multiaddr_to_socketaddr` fn to extract the SocketAddr

`multiaddr_to_socketaddr` cannot handle DNS

Reply:
is it possible to use `TcpTransport::new(timeout, None).listen(multiaddr)` directly in the network mod?

Reply:
`transports` is a private module, and it needs async runtime support，It is not currently available, but it seems to be possible to use DnsResolver directly

Reply:
resolve by https://github.com/nervosnetwork/ckb/pull/3231/commits/cbbb33b56a326be6c008e66529469216b767ae86